lib/pyspark_preprocesing.py

```diff
@@ -52,3 +52,62 @@ def one_hot_encoding(
         df_ohe = df_ohe.drop(*columns)
 
     return df_ohe
+
+
+def normalize_by_group(
+    spark: SparkSession,
+    df: DataFrame,
+    group_by: List[str],
+    normalize: List[str],
+    drop_cols: bool = True,
+) -> DataFrame:
+    """Funkcja zwraca sparkowy DataFrame ze znormalizowanymi kolumnami według grup.
+    Dodane kolumny posiadają przedrostki 'norm_'. Normalizację dokonujemy przez odjęcie
+    średniej i podzielenie przez odchylenie standardowe obliczane wewnątrz każdej grupy.
+
+    :param spark:           pyspark.sql.SparkSession
+    :param df:              pyspark.sql.DataFrame
+    :param group_by:        lista kolumn według których dokonamy grupowania
+    :param normalize:       lista kolumn z wartościami do znormalizowania
+    :param drop_cols:       usunięcie oryginalnych kolumn
+    :return:                pyspark.sql.DataFrame"""
+
+    assert all(x in df.columns for x in group_by)
+    assert all(x in df.columns for x in normalize)
+    assert all(x not in df.columns for x in ["avg_" + x for x in normalize])
+    assert all(x not in df.columns for x in ["stdev_" + x for x in normalize])
+
+    df.createOrReplaceTempView("tmp_view")
+    select_col = group_by + [
+        "stddev(" + x + ") as stdev_" + x + ", avg(" + x + ") as avg_" + x
```

Comment:
nie jestem fanem robienia tego za pomocą sql-a ale zostawiłem tę możliwość i masz prawo do tego wyboru. Ale rozważ chociaż użycie f-stringów by czytelniej zapisać swój kod
```suggestion
       f'stddev({x}) as stdev_{x}, avg({x}) as avg_{x}' 
```

Reply:
Chciałem żeby ta funkcja była jak najbardziej uniwersalna a zrobienie tego w sql wydało mi się prostszym rozwiązaniem. Dodałem f-stringi tam gdzie było można